### ISSUE TITLE
Problems with deleting files via dataset page

### ISSUE THREAD

**Issue by landreev** [IQSS/dataverse#4573]
This *may* be high priority/important/scary (but needs to be tested and confirmed): 

We may have a problem with file delete functionality, where a user attempts to delete a datafile, and while the physical file gets deleted from S3, the database objects (dvobject, datafile, filemetadata) are left behind. Any repeated attempts to delete such a datafile will fail - because the physical file is no longer there. 

I'll add more information as it becomes available. 

Case in point: support ticket RT 260355

**Comment by landreev** [comment]
In addition to the 5 files mentioned in RT 260355, there are 5 more missing files created since the migration, in 2 datasets. 

To be clear, some, or even all of these 10 missing files may still turn out to be botched *uploads*, and not botched deletes. 
As I said earlier, this still needs to be investigated and confirmed.

**Comment by landreev** [comment]
OK, here's a reproducible bug: 

TL;DR: once there's a "zombie" file in the dataset, trying to delete it can *add more zombie files* to the dataset. 

details: 

Once there is one missing S3 files in the dataset, attempting to delete it from the dataset page via the multiple file delete checkboxes, CAN RESULT IN MORE SUCH MISSING/ZOMBIE DATAFILES: 
To run this experiment, you need some way to monitor the S3 directory where the files for the dataset are stored. You can do it through the browser; or install the aws command line tool. (the latter will be used in the examples above). 

Configure the dataverse to use S3. Upload several files into a dataset. DO NOT PUBLISH - it must be a draft with no published version. 

Now, delete 1 of the files directly on S3 (without touching the datafile in the Dataverse database). 
Look up the storage identifier in the dvobject table, and delete it: 
(for example)
aws s3 rm s3://iqssqa/10.5072/FK2/JLYQP2/162cca642bc-22e07badc565

Verify that the file is gone; try to download it through the dataset page. You should be getting a 500 (I mean, you will be - ideally we should be giving a clean "file not found" page; this exception handling is handled in the issue #4578). 

Now select 3 "good" files via checkboxes; and then, as the last checkbox, select this file that you just deleted on S3. Now go to Edit Files->Delete. 

You should get the annoying/confusing "Success! Files could not be deleted!" message. You will notice that you still have the same number of Datafiles in the dataset (i.e., none were deleted). 

However, checking the S3 directory: 

aws s3 ls s3://iqssqa/10.5072/FK2/JLYQP2 --recursive

Will show that you now have not 1, but 4 files missing there!

I.e., the exception thrown when we attempted to delete that one non-existing physical file, halted the whole delete process - so the datafiles were not removed from the Dataverse db. However, since that was the last file, the first 3 had been successfully deleted already. So now we have 4 zombie files in the dataset. 

Clearly, we need to re-work the order in which we are doing these things - deleting the physical files and the entries in the database.

**Comment by landreev** [comment]
To emphasize, this scenario requires a missing/"zombie" to be already present in the dataset. There are indications that this is what happened in real life in RT 260355 (more files went missing as the user was trying to get rid of the first missing file). BUT we still have no explanation of how that first file got lost. My guess it is some form of a failed upload; but I haven't been able to reproduce it. 
(but that should be handled in another issue)

**Comment by pdurbin** [comment]
Related: #4564

**Comment by pdurbin** [comment]
I'm attempting to reproduce this bug by following the steps above but I'm having a little trouble and might benefit from @landreev looking over my shoulder.
